### ISSUE TITLE
no-lonely-if flags when the if is not so lonely (nested if..else in else)

### ISSUE THREAD

**Issue by codemakerlive** [eslint/eslint#3400]
eslint v.: 1.1.0
eslinrc contains: ```"no-lonely-if": 1```

The following gives "unexpected if as only statement in an else block (no-lonely-if)":

```js
if( something ) {
   //... handle it
} else { 
   // when in "not something" state...
   if( check_something_different ) {
      // ... 
   } else {  // that something different is not
      //...
   }
}
```

If the if in the else didn't have it's own else + nothing else, then fine (refactor!)... but it does so belongs without warning.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by dozoisch** [comment]
Shouldn't that be written as: 

```
if( something ) {
   //... handle it
} else if( check_something_different ) {
   // ... 
} else { // that something different is not
   //...
}
```

**Comment by codemakerlive** [comment]
Not in this case, I hoped the example would show that

if (something) {
&nbsp;&nbsp;&nbsp;//...
} else {
&nbsp;&nbsp;&nbsp;//... when in !something state, need to check something else
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// if that something else is true handle it one way
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// else handle it a different way
}

If you wanted to refactor that the way you suggest it's more like

if( something ) {
&nbsp;&nbsp;&nbsp;//... handle it
} else if( !something && something_else ) {
&nbsp;&nbsp;&nbsp;// ...
} else if( !something && !something_else ) {
&nbsp;&nbsp;&nbsp;//...
}

**Comment by codemakerlive** [comment]
I mean you're right, but is that really what the no-lonely-if is trying to accomplish...
Seems to me like it becomes not much of a refactor when the else has if...else inside.

**Comment by IanVS** [comment]
@codemakerlive, could you give a minimal example of working code and ESLint output that you don't agree with?

**Comment by codemakerlive** [comment]
I could but that's about as minimal as programming gets - can't see how an example is going to help? 

Seems like the simple rule no-lonely-if secretly wants to be no-nested-if-else

**Comment by IanVS** [comment]
Examples help us understand what you are requesting.  Maybe I'm just missing it, but I'm not sure what you're asking for.

**Comment by codemakerlive** [comment]
True usually, but in this case is just going to cloud over what this is about.

I see the following as useful (practically occuring when code has been manually refactored):

if( foo ) {
&nbsp;&nbsp;&nbsp;//...
} else {
&nbsp;&nbsp;&nbsp;if( bar ) {
&nbsp;&nbsp;&nbsp;&nbsp;// then lonely and really should be if...else
&nbsp;&nbsp;&nbsp;}
}

But this:

if( something ) {
&nbsp;&nbsp;&nbsp;//... handle it
} else if( check_something_different ) {
&nbsp;&nbsp;&nbsp;// ... 
} else { // that something different is not
&nbsp;&nbsp;&nbsp;//...
}

Is no better to quickly parse by eye and read than:

if( something ) {
&nbsp&nbsp;&nbsp;;//... handle it
} else {
&nbsp;&nbsp;&nbsp;// when in "not something" state...
&nbsp;&nbsp;&nbsp;if( check_something_different ) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// ...
&nbsp;&nbsp;&nbsp;} else { // that something different is not
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//...
&nbsp;&nbsp;&nbsp;}
}

It's subjective and trivial. But if you agree the rule is being overzealous then perhaps no-lonely-if can ignore if..else within a parent else block.

**Comment by michaelficarra** [comment]
@codemakerlive Just disable the rule. That's exactly the kind of thing this rule is intended to catch. Why do you want this rule turned on?

**Comment by nzakas** [comment]
@codemakerlive I agree with @michaelficarra here. You're asking us to add an exception to a rule whose sole purpose is to flag the thing you don't want to flag. :) In that case, it's best to disable the rule.
